File: sandbox/backend/java/simple-eccomerce/src/main/java/com/simple/ecommerce/infrastructure/persistence/InMemoryProductRepository.java
package com.simple.ecommerce.infrastructure.persistence;

import com.simple.ecommerce.application.exception.CategoryNotFoundException;
import com.simple.ecommerce.domain.entity.Category;
import com.simple.ecommerce.domain.entity.Product;
import com.simple.ecommerce.domain.repository.ProductRepository;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

/**
 * In-memory implementation of the ProductRepository interface.
 */
public class InMemoryProductRepository implements ProductRepository {

    private final List<Product> products;

    /**
     * Creates an empty in-memory product repository.
     */
    public InMemoryProductRepository() {
        this.products = new ArrayList<>();
    }

    @Override
    public void addProduct(Product product) {
        if (product == null) {
            throw new IllegalArgumentException("Product cannot be null");
        }
        products.add(product);
    }

    @Override
    public List<Product> getAllProducts() {
        return new ArrayList<>(products);
    }

    @Override
    public List<Product> getProductsSortedAlphabetically() {
        return products.stream()
                .sorted(Comparator.comparing(Product::getName))
                .collect(Collectors.toList());
    }

    @Override
    public List<Product> getAvailableProductsByCategory(Category category) {
        if (category == null) {
            throw new CategoryNotFoundException("Category cannot be null");
        }

        return products.stream()
                .filter(Product::isAvailable)
                .filter(p -> p.getCategory().equals(category))
                .sorted(Comparator.comparing(Product::getPrice))
                .collect(Collectors.toList());
    }
}

File: Projects/Testtrack/p911 - Simple eCommerce backend/infrastructure/persistence/in_memory_product_repository.py
from typing import List
from application import CategoryNotFoundException
from domain import Category, Product, ProductRepository


class InMemoryProductRepository(ProductRepository):
    def __init__(self):
        self.products: List[Product] = []

    def add_product(self, product: Product) -> None:
        if product is None:
            raise ValueError("Product cannot be null")
        else:
            self.products.append(product)

    def get_all_products(self) -> List[Product]:
        return self.products.copy()

    def get_products_sorted_alphabetically(self) -> List[Product]:
        return sorted(self.products, key=lambda p: p.get_name())

    def get_available_products_by_category(self, category: Category) -> List[Product]:
        if category is None:
            raise CategoryNotFoundException("Category cannot be null")

        available_products = [
            p for p in self.products if p.is_available() and p.get_category() == category]
        return sorted(available_products, key=lambda p: p.get_price())
